containers/message-parser/app/utils.py

```diff
@@ -378,153 +306,154 @@ def clean_schema(schema: dict):
+                reference_path = self._get_reference(field_parser, current_message)
+                value = fhirpathpy.evaluate(
+                    self.message,
+                    field_parser["fhir_path"],
+                    context={"ref": reference_path},
+                )  # Evaluate on full message, not current
+
+                if not value or len(value) == 0:
+                    return None
+                return value
+
+            if "fhir_path" in field_parser:
+                value = fhirpathpy.evaluate(current_message, field_parser["fhir_path"])
+
+            if not value or len(value) == 0:
+                return None
+
+            return value
+
+        # By default, fhirpathpy will compile such that *only*
+        # actual resources can be accessed, rather than data types.
+        # This is fine for most cases, but sometimes the actual data
+        # we want is in a list of structs rather than a list of
+        # resources, such as a list of patient addresses. This
+        # exception catches that and allows an ordinary property
+        # search.
+        except KeyError:
+            try:
+                accessors = field_parser["fhir_path"].split(".")[1:]
+                val = current_message
+                for acc in accessors:
+                    if "[" not in acc:
+                        val = val[acc]
+                    else:
+                        sub_acc = acc.split("[")[1].split("]")[0]
+                        val = val[acc.split("[")[0].strip()][int(sub_acc)]
+                return str(val)
+            except Exception:
+                return None
+
+    def _get_reference(self, field_parser, current_message):
+        """
+        Resolves a FHIR reference and returns a new path based on the resolved ID.
+
+        It uses a `reference_lookup` to find a reference ID in the `current_message`.
+        If a single reference is found, it uses the reference to replace a placeholder
+        in the `fhir_path` and evaluates this new path against the FHIR bundle
+        to return the referenced resource or value. If not, this function will raise
+        an error if those references cannot be resolved (if the ID of the referenced
+        object can't be found, for example).
+
+        :param field_parser: The parser for a specific field, which must contain a
+            `fhir_path` & a `reference_lookup`.
+        :param current_message: The FHIR message or sub-section at the current level of
+            parsing where the reference is located.
+        :return: The reference to be added as context to the FHIR path, or response
+            error message if the reference could not be resolved.
+        """
+        reference_parser = field_parser["reference_lookup"]
+        reference = fhirpathpy.evaluate(current_message, reference_parser)
+
+        if len(reference) == 0:
+            self.response.status_code = status.HTTP_400_BAD_REQUEST
+            return {

```

Comment:
why return and not throw? it looks like we're swallowing the details here vs using them anywhere